pkg/cli/admin/migrate/icsp/icsp.go

```diff
@@ -138,60 +141,82 @@ func (o *MigrateICSPOptions) ensureDirectoryViable() error {
+			return nil, fmt.Errorf("error converting to unstructured: %v", err)
+		}
+		delete(unstructuredObj.Object["metadata"].(map[string]interface{}), "creationTimestamp")
+		delete(unstructuredObj.Object, "status")
+		idmsBytes, err := yaml.Marshal(unstructuredObj.Object)
+		if err != nil {
+			return nil, fmt.Errorf("unable to marshal ImageDigestMirrorSet yaml: %v", err)
+		}
+		if len(icsps) > 1 && i != len(icsps)-1 {
+			idmsBytes = append(idmsBytes, []byte(yamlSeparator)...)
+		}
+		idmsYml = append(idmsYml, idmsBytes...)
 	}
 	return idmsYml, nil
 }
 
 // readICSPsFromFile appends to list of alternative image sources from ICSP file
 // returns error if no icsp object decoded from file data
-func readICSPsFromFile(icspFile string) (operatorv1alpha1.ImageContentSourcePolicy, string, error) {
+func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePolicy, string, error) {
 	icspData, err := os.ReadFile(icspFile)
 	if err != nil {
-		return operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("unable to read ImageContentSourceFile %s: %v", icspFile, err)
+		return []operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("unable to read ImageContentSourceFile %s: %v", icspFile, err)
 	}
 	if len(icspData) == 0 {
-		return operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("no data found in ImageContentSourceFile %s", icspFile)
-	}
-	icspObj, err := runtime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
-	if err != nil {
-		return operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", icspFile, err)
+		return []operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("no data found in ImageContentSourceFile %s", icspFile)
 	}
-	icsp, ok := icspObj.(*operatorv1alpha1.ImageContentSourcePolicy)
-	if !ok {
-		return operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("could not decode ImageContentSourcePolicy from %s", icspFile)
+
+	reader := utilyaml.NewYAMLReader(bufio.NewReader(bytes.NewReader(icspData)))
+	var icsps []operatorv1alpha1.ImageContentSourcePolicy
+	for {
+		icspBytes, err := reader.Read()
+		if err != nil && err != io.EOF {
+			return []operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("error reading ImageContentSourcePolicy from %s: %v", icspFile, err)
+		}
+		if icspBytes == nil {
+			break
+		}
+		icspObj, err := runtime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspBytes)
+		if err != nil {
+			return []operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", icspFile, err)
+		}
+		icsp, ok := icspObj.(*operatorv1alpha1.ImageContentSourcePolicy)
+		if !ok {
+			return []operatorv1alpha1.ImageContentSourcePolicy{}, "", fmt.Errorf("could not decode ImageContentSourcePolicy from %s", icspFile)
+		}
+		icsps = append(icsps, *icsp)
 	}
-	return *icsp, icsp.Name, nil
+	return icsps, icsps[0].Name, nil
```

Comment:
Who is preventing the panic in here when `icsps` is empty?

Reply:
In theory icsps is not empty when it returns from this line.
If the input file is empty, the "no data error" catches this error earlier https://github.com/openshift/oc/blob/7dfe72fee0e8d1ff8c4b7c82530311fd0b232e9f/pkg/cli/admin/migrate/icsp/icsp.go#L198-L200
If the input file has data the is not in the icsp scheme, the runtime decode can return the error https://github.com/openshift/oc/blob/7dfe72fee0e8d1ff8c4b7c82530311fd0b232e9f/pkg/cli/admin/migrate/icsp/icsp.go#L212-L215